packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1541,72 +1267,342 @@ module.exports = function<T, P, I, TI, PI, C, CX, PL>(
     return mostRecentCurrentTime;
   }
 
+  function deferredUpdates<A>(fn: () => A): A {
+    const previousExpirationContext = expirationContext;
+    expirationContext = computeAsyncExpiration();
+    try {
+      return fn();
+    } finally {
+      expirationContext = previousExpirationContext;
+    }
+  }
+
+  function syncUpdates<A>(fn: () => A): A {
+    const previousExpirationContext = expirationContext;
+    expirationContext = Sync;
+    try {
+      return fn();
+    } finally {
+      expirationContext = previousExpirationContext;
+    }
+  }
+
+  // TODO: Everything below this is written as if it has been lifted to the
+  // renderers. I'll do this in a follow-up.
+
+  // Represents a renderer root, e.g. the result of ReactDOM.createRoot
+  type ReactRoot = {|
+    +internalRoot: FiberRoot,
+    // Represents the work that on this root that needs to be flushed. To avoid
+    // leaking the fact that NoWork is a magic number, we use null to represent
+    // the absence of work.
+    remainingWork: ExpirationTime | null,
+    // True if the root is complete and it's ready to be committed. This flag
+    // is only true in between the render phase and commit phase. If we re-
+    // render a root without committing it, set it back to false. Attempting to
+    // commit a root that isn't ready will throw an error.
+    isReadyForCommit: boolean,
+  |};
+
+  // TODO: Is a map the right data structure here? We need to associate a
+  // an opqaue, internal root (FiberRoot) with a renderer root (what I'm
+  // calling ReactRoot here).
+  const scheduledRoots: Map<FiberRoot, ReactRoot> = new Map();
+  let isCallbackScheduled: boolean = false;
+  let isFlushingWork: boolean = false;
+  let nextFlushedRoot: ReactRoot | null = null;
+  let nextFlushedExpirationTime: ExpirationTime | null = null;
+  let deadlineDidExpire: boolean = false;
+  let hasUnhandledError: boolean = false;
+  let unhandledError: mixed | null = null;
+  let deadline: Deadline | null = null;
+
+  let isBatchingUpdates: boolean = false;
+  let isUnbatchingUpdates: boolean = false;
+
+  // Use these to prevent an infinite loop of nested updates
+  const NESTED_UPDATE_LIMIT = 1000;
+  let nestedUpdateCount: number = 0;
```

Comment:
Let's move this back up to the scheduler part somehow.

Reply:
The reason I moved this here is so that it's not coupled to the magic values for Sync and Task. The only reason it's still a global is so the error is thrown inside `setState`, for nice stack traces. Otherwise it could be local to the while loop in `flushWork`, which is where it conceptually belongs.